In [239]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sqlalchemy import create_engine

In [240]:
# load data as dataframe from sql
engine = create_engine('postgresql://postgres:postgres@localhost:5432/Nashville')
violation_types_df = pd.read_sql_query("SELECT * FROM violation_codes", engine)
# visually inspect
violation_types_df.head(5)

violation_id org_id violation_type                  violation_desc  \
0            1      1        BANNERS                         Banners   
1            2      1     BGMATOWNER  Proper Maintenance Req - Owner   
2            4      1      BLDGMAINT          Proper Maintenance Req   
3            5      1     BLDGPERMIT        Building Permit Required   
4            6      1      BLDGSCOPE          Scope of Building Code   

                                      violation_text     remedial_text  \
0  Section 17.32.060 & 17.32.070 -  Banners: Bann...      CAAZ_BANNERS   
1  Section 16.16.030 (B) - Proper maintenance req...  CAAH_BG_MATOWNER   
2  Section 16.16.030 (A) - Proper maintenance req...   CAAH_BLDG_MAINT   
3  Section 16.28.010 - Building Permit Required: ...  CAAB_BLDG_PERMIT   
4  Section 16.16.020 (B) - The provisions of this...   CAAH_BLDG_SCOPE   

  table_name expired_flag date_expired created_by date_created modified_by  \
0       NULL            N         NULL          1      03:52.6        NULL   
1       NULL            N         NULL          1      03:52.6        NULL   
2       NULL            N         NULL          1      03:52.6        NULL   
3       NULL            N         NULL          1      03:52.6        NULL   
4       NULL            N         NULL          1      03:52.6        NULL   

  date_modified fee_setup_id  
0          NULL         NULL  
1          NULL         NULL  
2          NULL         NULL  
3          NULL         NULL  
4          NULL         NULL

In [241]:
violation_types_df.tail(5)

violation_id org_id violation_type                    violation_desc  \
929          965      1          BBR36    No motions for reconsideration   
930          966      1          BBR37             Special Event Permits   
931          967      1          BBR38                  Suspension Rules   
932          969      1          BBR41  No permit issued with complaints   
933          970      1          BBR39       Sports Authority Facilities   

                                        violation_text remedial_text  \
929  The Beer Permit Board shall not entertain moti...          NULL   
930  Applicants for special event permits will be l...          NULL   
931  Upon suspension of a beer permit, the permitte...          NULL   
932  A permit shall not be issued by the Executive ...          NULL   
933  This rule is applicable only to beer sales wit...          NULL   

    table_name expired_flag date_expired created_by date_created modified_by  \
929       NULL            N         NULL      10635      04:56.0        NULL   
930       NULL            N         NULL      10635      08:59.4        NULL   
931       NULL            N         NULL      10635      10:16.9        NULL   
932       NULL            N         NULL      10635      12:53.1        NULL   
933       NULL            N         NULL      10635      16:46.8        NULL   

    date_modified fee_setup_id  
929          NULL         NULL  
930          NULL         NULL  
931          NULL         NULL  
932          NULL         NULL  
933          NULL         NULL

In [242]:
violation_types_df.columns

Index(['violation_id', 'org_id', 'violation_type', 'violation_desc',
       'violation_text', 'remedial_text', 'table_name', 'expired_flag',
       'date_expired', 'created_by', 'date_created', 'modified_by',
       'date_modified', 'fee_setup_id'],
      dtype='object')

In [243]:
# Ok. So, what I'm trying to do is reduce these 900ish to 20ish types. So let's decide which column we're going to reduce. 
# The metadata docx said the violation_id is a unique identifier. Let's test that.
print(violation_types_df.shape)

(934, 14)


In [244]:
violation_types_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 934 entries, 0 to 933
Data columns (total 14 columns):
violation_id      928 non-null object
org_id            925 non-null object
violation_type    925 non-null object
violation_desc    925 non-null object
violation_text    912 non-null object
remedial_text     788 non-null object
table_name        788 non-null object
expired_flag      788 non-null object
date_expired      788 non-null object
created_by        788 non-null object
date_created      651 non-null object
modified_by       651 non-null object
date_modified     651 non-null object
fee_setup_id      651 non-null object
dtypes: object(14)
memory usage: 102.2+ KB


In [245]:
# So, there are 934 rows, there should be 934 unique violation_id values. And, they should all be integers. 
# Let's see if they are unique.
violation_types_df.violation_id.value_counts(dropna = False)
#violation_types_df['violation_id'].value_counts(dropna = False)

-                                                                                                                                                                                                                                                                                                                     119
NaN                                                                                                                                                                                                                                                                                                                     6
75                                                                                                                                                                                                                                                                                                                      1
425                                                       

In [246]:
# Ok. So, 119 rows have a dash in them instead of an int. Let's look at them to seee if they are dropable
violation_types_df.query('violation_id == "-"').head(5)

violation_id org_id violation_type violation_desc violation_text  \
60             -   NULL           NULL              N           NULL   
155            -   NULL           NULL              N           NULL   
238            -   NULL           NULL              N           NULL   
322            -   NULL           NULL              N           NULL   
324            -   NULL           NULL              N           NULL   

    remedial_text table_name expired_flag date_expired created_by  \
60              2    00:00.0         NULL         NULL       NULL   
155             2    00:00.0         NULL         NULL       NULL   
238             2    00:00.0         NULL         NULL       NULL   
322             2    00:00.0         NULL         NULL       NULL   
324             2    00:00.0         NULL         NULL       NULL   

    date_created modified_by date_modified fee_setup_id  
60          None        None          None         None  
155         None        None          None         None  
238         None        None          None         None  
322         None        None          None         None  
324         None        None          None         None

In [247]:
violation_types_df.query('violation_id == "-"').tail(5)

violation_id org_id violation_type violation_desc violation_text  \
697            -   NULL           NULL              N           NULL   
699            -   NULL           NULL              N           NULL   
701            -   NULL           NULL              N           NULL   
703            -   NULL           NULL              N           NULL   
705            -   NULL           NULL              N           NULL   

    remedial_text table_name expired_flag date_expired created_by  \
697             2    00:00.0         NULL         NULL       NULL   
699             2    00:00.0         NULL         NULL       NULL   
701             2    00:00.0         NULL         NULL       NULL   
703             2    00:00.0         NULL         NULL       NULL   
705             2    00:00.0         NULL         NULL       NULL   

    date_created modified_by date_modified fee_setup_id  
697         None        None          None         None  
699         None        None          None         None  
701         None        None          None         None  
703         None        None          None         None  
705         None        None          None         None

## They definitely look dropable, but how do I confirm this programatically instead of relying on a visual inspection?
####      Well, I've confirmed that the head and tail match a pattern of "dropability" (i.e.:'-, NULL, NULL, N, etc...'). 
So, I could write a loop that looks inside each row, and confirms a pattern match for each cell in that row, if it's a match then I can either mark those rows for dropping, or just drop them directly.
     But all that seems complex. Instead of checking every category of every row, maybe I could just check a subset of 'critical categories' (save some time and effort). 
##### What are my best candidates for critical categories?
   If they don't have a type, description, or text, then they are effectively uncategorizeable. That's a nicer small set of 3. (Time data is irrelevant in this context since we're dealing with violation types, not actual instances of violations.) Ok so I'll try to write a loop that checks the values in those columns for each row.

In [248]:
# I should be able to query the dropable rows and store them as a new dataframe
dropabledf = violation_types_df.query('violation_type == "NULL" and violation_desc == "N" and violation_text == "NULL"')
dropabledf.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 137 entries, 15 to 926
Data columns (total 14 columns):
violation_id      131 non-null object
org_id            137 non-null object
violation_type    137 non-null object
violation_desc    137 non-null object
violation_text    137 non-null object
remedial_text     137 non-null object
table_name        137 non-null object
expired_flag      137 non-null object
date_expired      137 non-null object
created_by        137 non-null object
date_created      0 non-null object
modified_by       0 non-null object
date_modified     0 non-null object
fee_setup_id      0 non-null object
dtypes: object(14)
memory usage: 16.1+ KB


it's interesting that I ended up with 137. That's more than the 119 that have a dash instead of a violiation_id but that's fine because the criteria for discrimination applies to the entirety of the original df. If any row is missing those three critical categories then they are uncategorizeable. And that's my primary goal with this df; to reduce it from 900ish to 20ish.

## Ok, so now I've created a df where I can put everything I want to drop. How do I drop it from the original df? 
Some sort of merge? An outermerge should eliminate all entries from the original df that are in the dropable df, since it is a proper subset of the original df. This is a defacto solution. Let's try it.

In [249]:
#help(pd.merge) makes me think that a merge may not be the right approach.

In [250]:
#let's try df.replace instead

Ok, so I'm stuck. I can drop rows by index, but I have to pass the index values as a list. Can I just get the index list from the dropable df? Let's see

In [251]:
dropabledf.index

Int64Index([ 15,  60, 155, 238, 322, 324, 380, 400, 414, 416,
            ...
            708, 896, 903, 906, 908, 914, 917, 920, 923, 926],
           dtype='int64', length=137)

In [252]:
type(dropabledf.index)

pandas.core.indexes.numeric.Int64Index

In [253]:
cleaned_df1 = violation_types_df.drop(dropabledf.index)

In [254]:
cleaned_df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 797 entries, 0 to 933
Data columns (total 14 columns):
violation_id      797 non-null object
org_id            788 non-null object
violation_type    788 non-null object
violation_desc    788 non-null object
violation_text    775 non-null object
remedial_text     651 non-null object
table_name        651 non-null object
expired_flag      651 non-null object
date_expired      651 non-null object
created_by        651 non-null object
date_created      651 non-null object
modified_by       651 non-null object
date_modified     651 non-null object
fee_setup_id      651 non-null object
dtypes: object(14)
memory usage: 93.4+ KB


In [255]:
cleaned_df1['violation_id'].value_counts(dropna = False)

144                                                                                                                                                                        1
425                                                                                                                                                                        1
649                                                                                                                                                                        1
109                                                                                                                                                                        1
929                                                                                                                                                                        1
636                                                                                                                                    

In [256]:
# Ok. Let's see if I can make a sub df that contains all the rows where the violation id
# isn't an integer
# cleaned_df1.query(cleaned_df1.violation_id.is_integer()) didn't work

In [257]:
# cleaned_df1['violation_id'].apply(is_integer()) didn't work

In [258]:
# help(is_integer)

In [259]:
# help(is_int)

In [260]:
# help(pd.is_int)

In [261]:
# help(pd.is_integer)

In [262]:
# help(np.is_int)

In [263]:
# help(isinstance)

In [264]:
#cleaned_df1.query(cleaned_df1['violation_id'].apply(type == 'int'))

In [265]:
vid_series = cleaned_df1.violation_id

In [266]:
vid_series

0                                                      1
1                                                      2
2                                                      4
3                                                      5
4                                                      6
5                                                      7
6                                                      8
7                                                      9
8                                                     10
9                                                     11
10                                                    12
11                                                    13
12                                                    14
13                                                    15
14                                                    16
16                                                    17
17                                                    18
18                             

In [267]:
vid_series.apply(type)

0      <class 'str'>
1      <class 'str'>
2      <class 'str'>
3      <class 'str'>
4      <class 'str'>
5      <class 'str'>
6      <class 'str'>
7      <class 'str'>
8      <class 'str'>
9      <class 'str'>
10     <class 'str'>
11     <class 'str'>
12     <class 'str'>
13     <class 'str'>
14     <class 'str'>
16     <class 'str'>
17     <class 'str'>
18     <class 'str'>
19     <class 'str'>
20     <class 'str'>
21     <class 'str'>
22     <class 'str'>
23     <class 'str'>
24     <class 'str'>
25     <class 'str'>
26     <class 'str'>
27     <class 'str'>
28     <class 'str'>
29     <class 'str'>
30     <class 'str'>
           ...      
895    <class 'str'>
897    <class 'str'>
898    <class 'str'>
899    <class 'str'>
900    <class 'str'>
901    <class 'str'>
902    <class 'str'>
904    <class 'str'>
905    <class 'str'>
907    <class 'str'>
909    <class 'str'>
910    <class 'str'>
911    <class 'str'>
912    <class 'str'>
913    <class 'str'>
915    <class 'str'>
916    <class

In [268]:
np.sort(vid_series)

array(['1', '10', '100', '101', '102', '103', '104', '105', '106', '107',
       '108', '109', '11', '110', '111', '112', '113', '114', '115', '116',
       '117', '118', '119', '12', '120', '121', '122', '123', '124', '125',
       '126', '127', '128', '129', '13', '130', '131', '132', '133', '134',
       '135', '136', '137', '138', '139', '14', '140', '141', '142', '143',
       '144', '145', '146', '147', '148', '149', '15', '150', '151', '152',
       '153', '154', '155', '156', '157', '158', '159', '16', '160', '161',
       '162', '163', '164', '165', '166', '167', '168', '169', '17', '170',
       '171', '172', '173', '174', '175', '176', '177', '178', '179', '18',
       '180', '181', '182', '183', '184', '185', '186', '187', '188',
       '189', '19', '190', '191', '192', '193', '194', '195', '196', '197',
       '198', '199', '2', '20', '200', '201', '202', '203', '204', '205',
       '206', '207', '208', '209', '21', '210', '212', '213', '214', '22',
       '23', '24', '247

In [269]:
#vid_series.query('len =< 4') This didn't work

In [270]:
#cleaned_df1['violation_id'].query(len < 4) this also didn't work

In [271]:
#cleaned_df1.query('violation_id.len == 4') also didn't work

In [272]:
#vids_numeric = vid_series.apply(pd.to_numeric(errors = 'coerce')) also didn't work

In [273]:
vids_numeric = pd.to_numeric(vid_series, errors = 'coerce', downcast='integer')

In [274]:
vids_numeric

0        1.0
1        2.0
2        4.0
3        5.0
4        6.0
5        7.0
6        8.0
7        9.0
8       10.0
9       11.0
10      12.0
11      13.0
12      14.0
13      15.0
14      16.0
16      17.0
17      18.0
18      19.0
19      20.0
20      21.0
21      22.0
22      23.0
23      24.0
24      25.0
25      26.0
26      27.0
27      28.0
28      29.0
29      30.0
30      31.0
       ...  
895    946.0
897    947.0
898      NaN
899      NaN
900      NaN
901      NaN
902      NaN
904    948.0
905    949.0
907    951.0
909    952.0
910    953.0
911    954.0
912    955.0
913      NaN
915    956.0
916    957.0
918    959.0
919      NaN
921    960.0
922    968.0
924    961.0
925    962.0
927    963.0
928    964.0
929    965.0
930    966.0
931    967.0
932    969.0
933    970.0
Name: violation_id, Length: 797, dtype: float64

In [275]:
cleaned_df1.remedial_text.value_counts()

NULL                    443
CAAH_BLDG_SCOPE           2
CAAE_ELECEA_RECPT         1
CAAH_INT_INTDOOR          1
CAAP_PLUM_SYSHAZ          1
CAMCL_LIT_BUSREP          1
CAMCL_SEW_RUNOFF          1
CAAH_EXTER_MOCCUP         1
CAAH_EXT_DOOR             1
CAAP_WATER_WTRHEAT        1
CAMCL_LIT_PROPPL          1
CAAH_OCCUP_WCACC          1
CAAH_EXT_INSSCNRES        1
CAAH_EXT_TRLS             1
CAMCL_LIT_VEH             1
CAAH_BOARDING_REQU        1
CAAH_EXT_CHIMTW           1
CAAB_STOP_WORK            1
CAAH_INT_BLDSUR           1
CAAZ_SIGN_UNSAILL         1
CAAH_EXT_VEHRP            1
CAAH_EXT_WEED             1
CAAH_EXTER_REQ            1
CAAH_EXT_FDWALL           1
CAAH_EXT_FENMAT           1
CAAH_VENT_DRYER           1
CAAZ_CAR_ROW              1
CAAE_ELECFAC              1
CAAH_TRUCK                1
CAAP_PLUM_HOTEL           1
                       ... 
CAAZ_OVCROWDEDFM          1
CAAH_EXT_PTSID            1
CAAP_TOLRN_LOCEMFC        1
CAAE_ELECEQ_LIGHT         1
CAAH_OCCUP_RMWID    

In [276]:
def conv(x):
    try:
        int(x)
        return True
    except:
        return False
    
numeric = cleaned_df1['violation_id'].apply(conv)
cleaned_df2 = cleaned_df1[numeric]

In [277]:
cleaned_df2

violation_id org_id violation_type  \
0              1      1        BANNERS   
1              2      1     BGMATOWNER   
2              4      1      BLDGMAINT   
3              5      1     BLDGPERMIT   
4              6      1      BLDGSCOPE   
5              7      1     BOARDINGRQ   
6              8      1       BOARDVAC   
7              9      1         CARROW   
8             10      1       CERTCOMP   
9             11      1        CERTOCC   
10            12      1      CHICKFARM   
11            13      1     CONSIGSALE   
12            14      1      CROWDEDFM   
13            15      1        CROWDGH   
14            16      1      DUMPDESIG   
16            17      1       DUMPSTER   
17            18      1        ECSYSHZ   
18            19      1     ECSYSHZCRT   
19            20      1      ELECEAREC   
20            21      1         ELECEQ   
21            22      1     ELECEQLIGH   
22            23      1        ELECFAC   
23            24      1     ELECSVCPAN   
24            25      1         EXPARK   
25            26      1         EXTACC   
26            27      1       EXTBAHAT   
27            28      1         EXTBAL   
28            29      1      EXTCHIMTW   
29            30      1         EXTDEC   
30            31      1        EXTDECF   
..           ...    ...            ...   
888          939      1           BBR6   
889          940      1           BBR8   
890          941      1           BBR9   
891          942      1          BBR10   
892          943      1          BBR12   
893          944      1          BBR13   
894          945      1         BBR14A   
895          946      1         BBR14B   
897          947      1          BBR15   
904          948      1          BBR16   
905          949      1          BBR17   
907          951      1          BBR19   
909          952      1          BBR21   
910          953      1          BBR22   
911          954      1          BBR24   
912          955      1          BBR26   
915          956      1          BBR28   
916          957      1          BBR29   
918          959      1          BBR30   
921          960      1          BBR31   
922          968      1          BBR40   
924          961      1          BBR32   
925          962      1          BBR34   
927          963      1          BBR33   
928          964      1          BBR35   
929          965      1          BBR36   
930          966      1          BBR37   
931          967      1          BBR38   
932          969      1          BBR41   
933          970      1          BBR39   

                               violation_desc  \
0                                     Banners   
1              Proper Maintenance Req - Owner   
2                      Proper Maintenance Req   
3                    Building Permit Required   
4                      Scope of Building Code   
5                       Boarding Requirements   
6                Boarding of Vacant Buildings   
7                 Vehicle In The Right-Of-Way   
8          Certificate of Compliance Required   
9           Certificate of Occupancy Required   
10   Chickens and Other Domestic Farm Animals   
11                           Consignment Sale   
12            Overcrowding – Unrelated Adults   
13                  Overcrowding – Group Home   
14                Dumping in permitted places   
16    Hours of Collection & Dumpster Required   
17                  Electrical System Hazards   
18            Electrical System Certification   
19                     Electrical Receptacles   
20                      Electrical Equipment    
21               Electrical Lighting Fixtures   
22             Electrical Facilities Required   
23                   Electrical Service Panel   
24    Exterior Prop Area – Vehicles – Parking   
25      Exterior Prop Area – Acces Structures   
26     Exterior of Bldgs – Basement Hatchways   
27          Exterior of Buildings - Balconies   
28    Exterior of Bldgs – C

In [278]:
cleaned_df2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 788 entries, 0 to 933
Data columns (total 14 columns):
violation_id      788 non-null object
org_id            788 non-null object
violation_type    788 non-null object
violation_desc    788 non-null object
violation_text    775 non-null object
remedial_text     651 non-null object
table_name        651 non-null object
expired_flag      651 non-null object
date_expired      651 non-null object
created_by        651 non-null object
date_created      651 non-null object
modified_by       651 non-null object
date_modified     651 non-null object
fee_setup_id      651 non-null object
dtypes: object(14)
memory usage: 92.3+ KB


In [279]:
cleaned_df2.violation_id.value_counts(dropna=False).head(5)

144    1
412    1
953    1
551    1
852    1
Name: violation_id, dtype: int64

In [280]:
def violation_text_startswith(s):
    if s == None:
        return 'None'
    elif s.startswith('T.C.A') or s.startswith ('TCA'):
        return 'TCA'
    elif s.startswith('M.C.L') or s.startswith ('MCL'):
        return 'MCL'
    elif s.startswith('Section') or s.startswith ('section'):
        return 'SEC'
    else:
        return 'Other'

In [281]:
cleaned_df2['text_type'] = cleaned_df2['violation_text'].apply(violation_text_startswith)

/Users/paulomartinez/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [282]:
cleaned_df2

violation_id org_id violation_type  \
0              1      1        BANNERS   
1              2      1     BGMATOWNER   
2              4      1      BLDGMAINT   
3              5      1     BLDGPERMIT   
4              6      1      BLDGSCOPE   
5              7      1     BOARDINGRQ   
6              8      1       BOARDVAC   
7              9      1         CARROW   
8             10      1       CERTCOMP   
9             11      1        CERTOCC   
10            12      1      CHICKFARM   
11            13      1     CONSIGSALE   
12            14      1      CROWDEDFM   
13            15      1        CROWDGH   
14            16      1      DUMPDESIG   
16            17      1       DUMPSTER   
17            18      1        ECSYSHZ   
18            19      1     ECSYSHZCRT   
19            20      1      ELECEAREC   
20            21      1         ELECEQ   
21            22      1     ELECEQLIGH   
22            23      1        ELECFAC   
23            24      1     ELECSVCPAN   
24            25      1         EXPARK   
25            26      1         EXTACC   
26            27      1       EXTBAHAT   
27            28      1         EXTBAL   
28            29      1      EXTCHIMTW   
29            30      1         EXTDEC   
30            31      1        EXTDECF   
..           ...    ...            ...   
888          939      1           BBR6   
889          940      1           BBR8   
890          941      1           BBR9   
891          942      1          BBR10   
892          943      1          BBR12   
893          944      1          BBR13   
894          945      1         BBR14A   
895          946      1         BBR14B   
897          947      1          BBR15   
904          948      1          BBR16   
905          949      1          BBR17   
907          951      1          BBR19   
909          952      1          BBR21   
910          953      1          BBR22   
911          954      1          BBR24   
912          955      1          BBR26   
915          956      1          BBR28   
916          957      1          BBR29   
918          959      1          BBR30   
921          960      1          BBR31   
922          968      1          BBR40   
924          961      1          BBR32   
925          962      1          BBR34   
927          963      1          BBR33   
928          964      1          BBR35   
929          965      1          BBR36   
930          966      1          BBR37   
931          967      1          BBR38   
932          969      1          BBR41   
933          970      1          BBR39   

                               violation_desc  \
0                                     Banners   
1              Proper Maintenance Req - Owner   
2                      Proper Maintenance Req   
3                    Building Permit Required   
4                      Scope of Building Code   
5                       Boarding Requirements   
6                Boarding of Vacant Buildings   
7                 Vehicle In The Right-Of-Way   
8          Certificate of Compliance Required   
9           Certificate of Occupancy Required   
10   Chickens and Other Domestic Farm Animals   
11                           Consignment Sale   
12            Overcrowding – Unrelated Adults   
13                  Overcrowding – Group Home   
14                Dumping in permitted places   
16    Hours of Collection & Dumpster Required   
17                  Electrical System Hazards   
18            Electrical System Certification   
19                     Electrical Receptacles   
20                      Electrical Equipment    
21               Electrical Lighting Fixtures   
22             Electrical Facilities Required   
23                   Electrical Service Panel   
24    Exterior Prop Area – Vehicles – Parking   
25      Exterior Prop Area – Acces Structures   
26     Exterior of Bldgs – Basement Hatchways   
27          Exterior of Buildings - Balconies   
28    Exterior of Bldgs – C

In [283]:
cleaned_df2.query('text_type == "TCA"').info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 72 entries, 807 to 882
Data columns (total 15 columns):
violation_id      72 non-null object
org_id            72 non-null object
violation_type    72 non-null object
violation_desc    72 non-null object
violation_text    72 non-null object
remedial_text     72 non-null object
table_name        72 non-null object
expired_flag      72 non-null object
date_expired      72 non-null object
created_by        72 non-null object
date_created      72 non-null object
modified_by       72 non-null object
date_modified     72 non-null object
fee_setup_id      72 non-null object
text_type         72 non-null object
dtypes: object(15)
memory usage: 9.0+ KB


In [284]:
cleaned_df2.query('text_type == "TCA"').violation_id.value_counts(dropna = False)

855    1
882    1
887    1
900    1
929    1
883    1
876    1
897    1
903    1
866    1
910    1
892    1
863    1
899    1
909    1
868    1
920    1
871    1
881    1
916    1
870    1
906    1
859    1
875    1
858    1
908    1
911    1
857    1
886    1
919    1
      ..
854    1
891    1
927    1
934    1
893    1
904    1
856    1
898    1
896    1
864    1
880    1
888    1
885    1
879    1
917    1
878    1
889    1
894    1
914    1
853    1
907    1
915    1
861    1
890    1
877    1
902    1
905    1
895    1
873    1
913    1
Name: violation_id, Length: 72, dtype: int64

In [285]:
violations_TCA = cleaned_df2.query('text_type == "TCA"')

In [286]:
violations_TCA.head(72)

violation_id org_id violation_type  \
807          896      1       BT575416   
808          853      1     BT575101A2   
809          854      1     BT575101A3   
810          855      1     BT575103A1   
811          856      1    BT575103A10   
812          857      1     BT575103A2   
813          858      1    BT575103A3A   
814          859      1    BT575103A3B   
815          860      1    BT575103A3C   
816          861      1     BT575103A4   
817          862      1     BT575103A5   
818          863      1     BT575103A6   
819          864      1     BT575103D1   
820          865      1     BT575103D2   
821          866      1     BT575105B1   
822          867      1     BT575105B2   
823          868      1     BT575105B3   
824          897      1     BT5755031A   
825          869      1     BT575105B4   
826          870      1     BT575105B5   
827          871      1      BT575105D   
828          872      1      BT575107A   
829          873      1      BT575107B   
830          874      1      BT575110A   
831          875      1      BT575110B   
832          876      1     BT575110E2   
833          877      1    BT575111B1A   
834          878      1    BT575111B1B   
835          879      1    BT575111B1C   
836          880      1     BT575111B3   
..           ...    ...            ...   
849          892      1     BT575301B1   
850          893      1      BT575301C   
851          894      1      BT575301F   
852          898      1     BT5755031B   
853          899      1     BT5755032A   
854          900      1     BT5755032B   
855          901      1     BT5755032C   
856          902      1     BT5755032D   
857          903      1     BT5755032E   
858          904      1     BT5755032F   
859          905      1     BT5755032G   
860          906      1     BT5755032H   
861          907      1     BT5755032I   
862          908      1     BT5755032J   
863          909      1      BT575504A   
864          910      1      BT575504B   
865          911      1      BT575504C   
866          912      1     BT575504C1   
867          913      1     BT575504C2   
868          914      1     BT575504C3   
869          915      1     BT575504C4   
870          916      1      BT575504D   
871          917      1      BT575504E   
872          918      1      BT575608A   
873          919      1      BT575608B   
874          920      1      BT575608C   
875          927      1     BT5451083C   
876          928      1    BT574203B1A   
877          929      1      BT575409A   
882          934      1      BT5751038   

                               violation_desc  \
807              Beer stored at other address   
808     Not Manufacturer and Wholesale Retail   
809     Not Wholesale and Manufacturer Retail   
810      Operating without county city permit   
811         Permits to citizen legal resident   
812                  Permit to business owner   
813                 Cannot transfer ownership   
814           Permit for single location only   
815         Business Name same as Permit Name   
816    Two businesses one location one permit   
817               On Sale Off Sale one permit   
818      Return within 15 days of termination   
819  No valid cert to sell, dist, manufacture   
820     Resale cert not displayed distributed   
821              Sales near schools, churches   
822                        No sales to minors   
823   Law violations by owners in past 10 yrs   
824        Wholesaler transfer written notice   
825    Employee law violations in past 10 yrs   
826  Onsale permit for on premise consumption   
827            False statement on application   
828          Hotel motel locked in room units   
829          Hotel motel no under 21 get keys   
830                             Bond required   
831                       Bond rider required   
832                           Bond revocation   
833                Homemade Beer Compensation   
834                       H

I expected to see more TCA sections than just 57, check to see if they're there.

In [287]:
sorted(violations_TCA.violation_text)

['T.C.A 57-4-203(B)(1)(A): any licensee or other person who sells, furnishes, disposes of, gives, or causes to be sold, furnished, disposed of, or given, any alcoholic beverage to any person under 21 years of age',
 'T.C.A 57-5-101(a)(2): Except as otherwise provided for in this part, no brewer or manufacturer of beer shall have any financial or ownership interest, direct or indirect, in the business of or a building containing a wholesale or retail licensee, including to furnish or loan any fixtures of any kind to a retail licensee, and no such brewer or manufacturer shall hold a wholesale or retail license. For purposes of this section, an indirect interest includes any interest acquired by affiliates, subsidiaries, corporate officials, partners, or employees of the brewer or manufacturer.',
 'T.C.A 57-5-101(a)(3): Except as otherwise provided for in this part, no wholesaler shall hold any financial or ownership interest, direct or indirect, in the business of or a building containin

They're no.
So, did something went wrong with my labeling strategy?

In [288]:
cleaned_df2.text_type.value_counts(dropna = False)

SEC      429
Other    173
MCL      101
TCA       72
None      13
Name: text_type, dtype: int64

In [289]:
print("the counta of the text types is ", 429+173+101+72+13)
cleaned_df2.shape

the counta of the text types is  788


(788, 15)

In [290]:
def TCA_text_startswith(s):
    if s == None:
        return 'None'
    elif s.startswith('T.C.A 57-4'):
        return '57-4, Intoxicating Liquors - Consumption of Alcoholic Beverages on Premises'
    elif s.startswith('T.C.A 57-5'):
        return '57-5, Intoxicating Liquors - Beer'
    else:
        return 'Other'

In [291]:
cleaned_df2['text_subtype_and_text_subsubtype'] = cleaned_df2['violation_text'].apply(TCA_text_startswith)

/Users/paulomartinez/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [292]:
cleaned_df2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 788 entries, 0 to 933
Data columns (total 16 columns):
violation_id                        788 non-null object
org_id                              788 non-null object
violation_type                      788 non-null object
violation_desc                      788 non-null object
violation_text                      775 non-null object
remedial_text                       651 non-null object
table_name                          651 non-null object
expired_flag                        651 non-null object
date_expired                        651 non-null object
created_by                          651 non-null object
date_created                        651 non-null object
modified_by                         651 non-null object
date_modified                       651 non-null object
fee_setup_id                        651 non-null object
text_type                           788 non-null object
text_subtype_and_text_subsubtype    788 non-null object

Ok. I think I've successfully given meaningful categories to everything that is TCA. Lets look at it as a horizontal slice.

In [293]:
cleaned_df2.query('text_type == "TCA"')

violation_id org_id violation_type  \
807          896      1       BT575416   
808          853      1     BT575101A2   
809          854      1     BT575101A3   
810          855      1     BT575103A1   
811          856      1    BT575103A10   
812          857      1     BT575103A2   
813          858      1    BT575103A3A   
814          859      1    BT575103A3B   
815          860      1    BT575103A3C   
816          861      1     BT575103A4   
817          862      1     BT575103A5   
818          863      1     BT575103A6   
819          864      1     BT575103D1   
820          865      1     BT575103D2   
821          866      1     BT575105B1   
822          867      1     BT575105B2   
823          868      1     BT575105B3   
824          897      1     BT5755031A   
825          869      1     BT575105B4   
826          870      1     BT575105B5   
827          871      1      BT575105D   
828          872      1      BT575107A   
829          873      1      BT575107B   
830          874      1      BT575110A   
831          875      1      BT575110B   
832          876      1     BT575110E2   
833          877      1    BT575111B1A   
834          878      1    BT575111B1B   
835          879      1    BT575111B1C   
836          880      1     BT575111B3   
..           ...    ...            ...   
849          892      1     BT575301B1   
850          893      1      BT575301C   
851          894      1      BT575301F   
852          898      1     BT5755031B   
853          899      1     BT5755032A   
854          900      1     BT5755032B   
855          901      1     BT5755032C   
856          902      1     BT5755032D   
857          903      1     BT5755032E   
858          904      1     BT5755032F   
859          905      1     BT5755032G   
860          906      1     BT5755032H   
861          907      1     BT5755032I   
862          908      1     BT5755032J   
863          909      1      BT575504A   
864          910      1      BT575504B   
865          911      1      BT575504C   
866          912      1     BT575504C1   
867          913      1     BT575504C2   
868          914      1     BT575504C3   
869          915      1     BT575504C4   
870          916      1      BT575504D   
871          917      1      BT575504E   
872          918      1      BT575608A   
873          919      1      BT575608B   
874          920      1      BT575608C   
875          927      1     BT5451083C   
876          928      1    BT574203B1A   
877          929      1      BT575409A   
882          934      1      BT5751038   

                               violation_desc  \
807              Beer stored at other address   
808     Not Manufacturer and Wholesale Retail   
809     Not Wholesale and Manufacturer Retail   
810      Operating without county city permit   
811         Permits to citizen legal resident   
812                  Permit to business owner   
813                 Cannot transfer ownership   
814           Permit for single location only   
815         Business Name same as Permit Name   
816    Two businesses one location one permit   
817               On Sale Off Sale one permit   
818      Return within 15 days of termination   
819  No valid cert to sell, dist, manufacture   
820     Resale cert not displayed distributed   
821              Sales near schools, churches   
822                        No sales to minors   
823   Law violations by owners in past 10 yrs   
824        Wholesaler transfer written notice   
825    Employee law violations in past 10 yrs   
826  Onsale permit for on premise consumption   
827            False statement on application   
828          Hotel motel locked in room units   
829          Hotel motel no under 21 get keys   
830                             Bond required   
831                       Bond rider required   
832                           Bond revocation   
833                Homemade Beer Compensation   
834                       H

In [294]:
cleaned_df2.query('text_type == "TCA"').info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 72 entries, 807 to 882
Data columns (total 16 columns):
violation_id                        72 non-null object
org_id                              72 non-null object
violation_type                      72 non-null object
violation_desc                      72 non-null object
violation_text                      72 non-null object
remedial_text                       72 non-null object
table_name                          72 non-null object
expired_flag                        72 non-null object
date_expired                        72 non-null object
created_by                          72 non-null object
date_created                        72 non-null object
modified_by                         72 non-null object
date_modified                       72 non-null object
fee_setup_id                        72 non-null object
text_type                           72 non-null object
text_subtype_and_text_subsubtype    72 non-null object
dtypes: object

In [295]:
cleaned_df2.text_subtype_and_text_subsubtype.value_counts(dropna = False)

Other                                                                          703
57-5, Intoxicating Liquors - Beer                                               71
None                                                                            13
57-4, Intoxicating Liquors - Consumption of Alcoholic Beverages on Premises      1
Name: text_subtype_and_text_subsubtype, dtype: int64

That's interesting. Why are there 13 nones. Let's look at them as text subtypes

In [296]:
cleaned_df2.text_type.value_counts(dropna=False)

SEC      429
Other    173
MCL      101
TCA       72
None      13
Name: text_type, dtype: int64

They're still there, which means that the violation text was probably none for 13 rows. Lets take a look at those rows.

In [297]:
cleaned_df2.query('text_type == "None"')

violation_id org_id violation_type  \
219          247      1            917   
710          756      1      HD1032120   
711          757      1      HD1032130   
712          758      1      HD1032140   
713          759      1      HD1032160   
714          760      1     HD10360201   
715          761      1     HD10360202   
716          762      1     HD10360203   
717          763      1     HD10360204   
718          764      1      HD1572020   
719          765      1       HD157240   
720          766      1      HD1624110   
725          771      1         HDVIOL   

                             violation_desc violation_text remedial_text  \
219                         HOUSE DISREPAIR           None          NULL   
710     APPROVED GARBAGE & TRASH CONTAINERS           None          NULL   
711         ACCUMUALTION OF GARBAGE & TRASH           None          NULL   
712  ACCUMUALTION OF LUMBER, BOXES, BARRELS           None          NULL   
713               STAGNANT WATER COLLECTION           None          NULL   
714          DIRT, FILTH, RUBBISH & GARBAGE           None          NULL   
715           VERMIN AND RODENT INFESTATION           None          NULL   
716                       EXCESS VEGETATION           None          NULL   
717                      CLEAN AMINAL AREAS           None          NULL   
718                        SEWAGE ON GROUND           None          NULL   
719   IMPROPER MAINT SEWAGE DISPOSAL SYSTEM           None          NULL   
720                         FAULTY PLUMBING           None          NULL   
725                    VEGETATION VIOLATION           None          NULL   

    table_name expired_flag date_expired created_by date_created modified_by  \
219       NULL            N         NULL          2      00:00.0        NULL   
710       NULL            N         NULL          2      00:00.0        NULL   
711       NULL            N         NULL          2      00:00.0        NULL   
712       NULL            N         NULL          2      00:00.0        NULL   
713       NULL            N         NULL          2      00:00.0        NULL   
714       NULL            N         NULL          2      00:00.0        NULL   
715       NULL            N         NULL          2      00:00.0        NULL   
716       NULL            N         NULL          2      00:00.0        NULL   
717       NULL            N         NULL          2      00:00.0        NULL   
718       NULL            N         NULL          2      00:00.0        NULL   
719       NULL            N         NULL          2      00:00.0        NULL   
720       NULL            N         NULL          2      00:00.0        NULL   
725       NULL            N         NULL          2      00:00.0        NULL   

    date_modified fee_setup_id text_type text_subtype_and_text_subsubtype  
219          NULL         NULL      None                             None  
710          NULL         NULL      None                             None  
711          NULL         NULL      None                             None  
712          NULL         NULL      None                             None  
713          NULL         NULL      None                             None  
714          NULL         NULL      None                             None  
715          NULL         NULL      None                             None  
716          NULL         NULL      None                             None  
717          NULL         NULL      None                             None  
718          NULL         NULL      None                             None  
719          NULL         NULL      None                             None  
720          NULL         NULL      None                             None  
725          NULL         NULL      None                             None

Weird. I'd drop them except that they seem to have meaningfull type and description... What I have noticed is that they all have types beginning with HD, and all the descriptions are germane to housing. Label their text_type as HD and look into them later

In [298]:
sorted(cleaned_df2.query('text_type == "None"').violation_type)

['917',
 'HD1032120',
 'HD1032130',
 'HD1032140',
 'HD1032160',
 'HD10360201',
 'HD10360202',
 'HD10360203',
 'HD10360204',
 'HD1572020',
 'HD157240',
 'HD1624110',
 'HDVIOL']

Yeah, definitely looks like HD is the text Type, 103,157,162 the sub types and the remaining numbe sequences indicative of the subsubtype. I guess write the relevant function. and categorize them.

Actually, before I do, I need to make sure that none of my other rows have HD types...

In [299]:
def violation_type_startswith(s):
    if s == None:
        return 'None'
    elif s.startswith('HD') or s.startswith('H.D') :
        return 'HD'
    elif s.startswith('BB') or s.startswith('B.B'):
        return 'BB'
    elif s.startswith('TCA') or s.startswith('T.C.A'):
        return 'TCA'
    elif s.startswith('MCL') or s.startswith('M.C.L'):
        return 'MCL'
    else:
        return 'Other'
cleaned_df2['violation_subtype'] = cleaned_df2['violation_type'].apply(violation_type_startswith)

/Users/paulomartinez/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:14: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [300]:
cleaned_df2.violation_subtype.value_counts(dropna = False)

Other    661
BB       111
HD        16
Name: violation_subtype, dtype: int64

That's interesting. There are 16 HD v_subtypes, that's 3 more than expected. Inspect

In [301]:
cleaned_df2.query('violation_subtype == "HD"')

violation_id org_id violation_type  \
710          756      1      HD1032120   
711          757      1      HD1032130   
712          758      1      HD1032140   
713          759      1      HD1032160   
714          760      1     HD10360201   
715          761      1     HD10360202   
716          762      1     HD10360203   
717          763      1     HD10360204   
718          764      1      HD1572020   
719          765      1       HD157240   
720          766      1      HD1624110   
721          767      1   HDPM_GT_RATS   
722          768      1       HDPM_JTD   
723          769      1      HDPM_STAG   
724          770      1      HDPM_VERM   
725          771      1         HDVIOL   

                             violation_desc  \
710     APPROVED GARBAGE & TRASH CONTAINERS   
711         ACCUMUALTION OF GARBAGE & TRASH   
712  ACCUMUALTION OF LUMBER, BOXES, BARRELS   
713               STAGNANT WATER COLLECTION   
714          DIRT, FILTH, RUBBISH & GARBAGE   
715           VERMIN AND RODENT INFESTATION   
716                       EXCESS VEGETATION   
717                      CLEAN AMINAL AREAS   
718                        SEWAGE ON GROUND   
719   IMPROPER MAINT SEWAGE DISPOSAL SYSTEM   
720                         FAULTY PLUMBING   
721                   RAT HARBORAGE GARBAGE   
722                JUNK TRASH RAT HARBORAGE   
723                          STAGNANT WATER   
724                                  VERMIN   
725                    VEGETATION VIOLATION   

                                        violation_text remedial_text  \
710                                               None          NULL   
711                                               None          NULL   
712                                               None          NULL   
713                                               None          NULL   
714                                               None          NULL   
715                                               None          NULL   
716                                               None          NULL   
717                                               None          NULL   
718                                               None          NULL   
719                                               None          NULL   
720                                               None          NULL   
721  MCL Chapter 10.32.130 – Permitting garbage and...          NULL   
722  MCL Chapter 10.32.140 - Permitting lumber, jun...          NULL   
723  MCL Chapter 10.32.160 – The chief medical dire...          NULL   
724  MCL Chapter 10.36.020 a -  Every dwelling and ...          NULL   
725                                               None          NULL   

    table_name expired_flag date_expired created_by date_created modified_by  \
710       NULL            N         NULL          2      00:00.0        NULL   
711       NULL            N         NULL          2      00:00.0        NULL   
712       NULL            N         NULL          2      00:00.0        NULL   
713       NULL            N         NULL          2      00:00.0        NULL   
714       NULL            N         NULL          2      00:00.0        NULL   
715       NULL            N         NULL          2      00:00.0        NULL   
716       NULL            N         NULL          2      00:00.0        NULL   
717       NULL            N         NULL          2      00:00.0        NULL   
718       NULL            N         NULL          2      00:00.0        NULL   
719       NULL            N         NULL          2      00:00.0        NULL   
720       NULL            N         NULL          2      00:00.0        NULL   
721       NULL            N         NULL          2      00:00.0        NULL   
722       NULL            N         NULL          2      00:00.0        NULL   
723       NULL            N         NULL          2      00:00.0        NULL   
724       NULL            N         NULL          2      00:00.0

alright so there seems to be a clue in the fact that they overlap with MCL. Hopefully that means that they are all from MCL. Anyway, that also means that I can keep going.

I bet the BB sub types are all MCL text types. Let's see how would I verify that?

In [302]:
cleaned_df2.query('violation_subtype == "BB"')

violation_id org_id violation_type  \
735          781      1       BB708020   
736          782      1      BB708030B   
737          783      1      BB708030C   
738          784      1      BB708030D   
739          785      1      BB708030E   
740          786      1     BB708040B3   
741          841      1      BB720010B   
742          787      1     BB708040B5   
743          788      1      BB708040G   
744          789      1      BB708040H   
745          790      1      BB708060B   
746          791      1      BB708070A   
747          792      1      BB708070B   
748          793      1      BB708070C   
749          794      1      BB708070D   
750          795      1       BB708080   
751          796      1       BB708100   
752          797      1      BB708105C   
753          798      1      BB708105D   
754          799      1      BB708105E   
755          800      1       BB708110   
756          801      1      BB708130A   
757          802      1      BB708130B   
758          803      1      BB708130C   
759          804      1      BB708140A   
760          805      1      BB708140B   
761          806      1      BB708140C   
762          807      1      BB708140D   
763          808      1      BB708140E   
764          809      1      BB708140F   
..           ...    ...            ...   
888          939      1           BBR6   
889          940      1           BBR8   
890          941      1           BBR9   
891          942      1          BBR10   
892          943      1          BBR12   
893          944      1          BBR13   
894          945      1         BBR14A   
895          946      1         BBR14B   
897          947      1          BBR15   
904          948      1          BBR16   
905          949      1          BBR17   
907          951      1          BBR19   
909          952      1          BBR21   
910          953      1          BBR22   
911          954      1          BBR24   
912          955      1          BBR26   
915          956      1          BBR28   
916          957      1          BBR29   
918          959      1          BBR30   
921          960      1          BBR31   
922          968      1          BBR40   
924          961      1          BBR32   
925          962      1          BBR34   
927          963      1          BBR33   
928          964      1          BBR35   
929          965      1          BBR36   
930          966      1          BBR37   
931          967      1          BBR38   
932          969      1          BBR41   
933          970      1          BBR39   

                               violation_desc  \
735                    Sales without a permit   
736    Off Sale - consume on or near premises   
737   On sale required for on premise consume   
738                            Caterer Permit   
739         Special Event - sales on premises   
740  Owners convict beer laws moral turpitude   
741               BB WMD contiguous locations   
742       Owners incarcerated moral turpitude   
743      Supplemental Application Information   
744            False statement on application   
745                  Payment of Privilege Tax   
746                           BB public place   
747                     BB invoices and memos   
748                         BB permit display   
749                BB Off Sale permit display   
750               BB permit address displayed   
751      On sale permit issuance requirements   
752         Caterer Advanced Notice of events   
753   Caterer location revoked on sale permit   
754   Caterer in 100 feet of church or school   
755                  Liquor sales to under 21   
756           BB prohibited drugs on premises   
757        BB beverage in excess of 8 percent   
758                    BB possession under 21   
759            BB Employ any person convicted   
760             BB Employ any person released   
761                 BB sell to under 21 years   
762                   BB se

In [303]:
cleaned_df2.query('violation_subtype == "BB"').info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 111 entries, 735 to 933
Data columns (total 17 columns):
violation_id                        111 non-null object
org_id                              111 non-null object
violation_type                      111 non-null object
violation_desc                      111 non-null object
violation_text                      111 non-null object
remedial_text                       102 non-null object
table_name                          102 non-null object
expired_flag                        102 non-null object
date_expired                        102 non-null object
created_by                          102 non-null object
date_created                        102 non-null object
modified_by                         102 non-null object
date_modified                       102 non-null object
fee_setup_id                        102 non-null object
text_type                           111 non-null object
text_subtype_and_text_subsubtype    111 non-null obje

In [304]:
cleaned_df2.query('violation_subtype == "BB"').text_type.value_counts(dropna = False)

MCL      76
Other    35
Name: text_type, dtype: int64

In [305]:
#weird... what are the ones with Other text type?

In [306]:
cleaned_df2.query('violation_subtype == "BB" and text_type == "Other"')

violation_id org_id violation_type  \
883          935      1          BBR11   
884          950      1          BBR18   
885          936      1           BBR1   
886          937      1           BBR2   
887          938      1           BBR3   
888          939      1           BBR6   
889          940      1           BBR8   
890          941      1           BBR9   
891          942      1          BBR10   
892          943      1          BBR12   
893          944      1          BBR13   
894          945      1         BBR14A   
895          946      1         BBR14B   
897          947      1          BBR15   
904          948      1          BBR16   
905          949      1          BBR17   
907          951      1          BBR19   
909          952      1          BBR21   
910          953      1          BBR22   
911          954      1          BBR24   
912          955      1          BBR26   
915          956      1          BBR28   
916          957      1          BBR29   
918          959      1          BBR30   
921          960      1          BBR31   
922          968      1          BBR40   
924          961      1          BBR32   
925          962      1          BBR34   
927          963      1          BBR33   
928          964      1          BBR35   
929          965      1          BBR36   
930          966      1          BBR37   
931          967      1          BBR38   
932          969      1          BBR41   
933          970      1          BBR39   

                               violation_desc  \
883                 Post laws and regulations   
884                         Health Department   
885                      Temp License Allowed   
886            Commercially sealed containers   
887        Under 18 transport, sell, dispense   
888                  No drive-through service   
889  Special Event bonafide charity political   
890                   All owners and partners   
891                               No delivery   
892                     Permittee responsible   
893               Lease or Ownership required   
894                    All applicants over 21   
895                          Ascertaining age   
897                   Special Event Approvals   
904                    Payment Types Accepted   
905                 Crimes of moral turpitude   
907                              Fire Marshal   
909                         Surrender permits   
910               Complaints filed in writing   
911                       Old or New Location   
912                Applications 10 days prior   
915       Cannot surrender to avoid penalties   
916  Cannot temporary permits with complaints   
918                             Grandfathered   
921                         Expanded premises   
922                Deny permit after 6 months   
924                           Caterer Permits   
925                        Standard penalties   
927                Past violations considered   
928                    LLC equals corporation   
929            No motions for reconsideration   
930                     Special Event Permits   
931                          Suspension Rules   
932          No permit issued with complaints   
933               Sports Authority Facilities   

                                        violation_text  \
883  All retail permittees shall conspicuously post...   
884  Upon making application for a beer permit from...   
885  The Executive Secretary to the Metropolitan Be...   
886  An off-sale beer permittee shall sell beer onl...   
887  No person under 18 years of age shall transpor...   
888  Beer shall not be sold through any drive-throu...   
889  Special Events held on public property can onl...   
890  All applicants for beer permits must supply th...   
891  No permittee nor any of his/her agents shall d...   
892  In disciplinary proceedings, it shall be no de...   
893  When applying for an on or off-sale permit, al...   
894  All applicants for a perm

In [307]:
cleaned_df2.query('violation_subtype == "BB" and text_type == "Other"').info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 35 entries, 883 to 933
Data columns (total 17 columns):
violation_id                        35 non-null object
org_id                              35 non-null object
violation_type                      35 non-null object
violation_desc                      35 non-null object
violation_text                      35 non-null object
remedial_text                       26 non-null object
table_name                          26 non-null object
expired_flag                        26 non-null object
date_expired                        26 non-null object
created_by                          26 non-null object
date_created                        26 non-null object
modified_by                         26 non-null object
date_modified                       26 non-null object
fee_setup_id                        26 non-null object
text_type                           35 non-null object
text_subtype_and_text_subsubtype    35 non-null object
violation_subt

Ok. So, I gave meaningful categories to everything with a text type TCA. Now I need to to the same for the other text types. What are they again?

In [313]:
cleaned_df2.text_type.value_counts(dropna = False)

SEC      429
Other    173
MCL      101
TCA       72
None      13
Name: text_type, dtype: int64

I'm pretty sure that I already looked into the 13 nones, and they turned out to be HD violation sub types (which in turn are likely MCL text types)